### ISSUE TITLE
Authentication fails from last connection

### ISSUE THREAD

**Issue by GDB4J** [Auties00/WhatsappWeb4j#150]
Hi @Auties00, 

I'm using this piece of code to get a connection.
When I removed all the serialized connections, I get a QR code and everything seems fine.
But when I restart my application, it's using the "lastConnection" and I got the error below.

Seems like something is off at serializing/deserializing the connections?

```
Whatsapp api;
if (Whatsapp.listConnections().isEmpty()) {
                var configuration = Whatsapp.Options.newOptions()
                        .description("MyApp")
                        .create();
                api = Whatsapp.newConnection(configuration);
            } else {
                api = Whatsapp.lastConnection();
            }
            api.connect().get();
```

Error
```
Caused by: java.lang.NullPointerException: Cannot invoke "it.auties.whatsapp.model.media.MediaConnection.auth()" because the return value of "it.auties.whatsapp.controller.Store.mediaConnection()" is null
	at it.auties.whatsapp.util.Medias.upload(Medias.java:51)
	at it.auties.whatsapp.model.message.standard.ImageMessage.simpleBuilder(ImageMessage.java:179)
	at it.auties.whatsapp.model.message.standard.ImageMessage$SimpleImageBuilder.create(ImageMessage.java:172)
	at be.geertdebaets.camera.service.WhatsAppService.sendImageMessageToGroup(WhatsAppService.java:147)
	... 1 more
```

Version: 
` <version>3.0-RC13</version>` (but was also on RC12)

Is this a bug/problem, or i'm a doing something wrong in setting up the connections?

**Comment by Auties00** [comment]
> Hi @Auties00,
> 
> I'm using this piece of code to get a connection. When I removed all the serialized connections, I get a QR code and everything seems fine. But when I restart my application, it's using the "lastConnection" and I got the error below.
> 
> Seems like something is off at serializing/deserializing the connections?
> 
> ```
> Whatsapp api;
> if (Whatsapp.listConnections().isEmpty()) {
>                 var configuration = Whatsapp.Options.newOptions()
>                         .description("MyApp")
>                         .create();
>                 api = Whatsapp.newConnection(configuration);
>             } else {
>                 api = Whatsapp.lastConnection();
>             }
>             api.connect().get();
> ```
> 
> Error
> 
> ```
> Caused by: java.lang.NullPointerException: Cannot invoke "it.auties.whatsapp.model.media.MediaConnection.auth()" because the return value of "it.auties.whatsapp.controller.Store.mediaConnection()" is null
> 	at it.auties.whatsapp.util.Medias.upload(Medias.java:51)
> 	at it.auties.whatsapp.model.message.standard.ImageMessage.simpleBuilder(ImageMessage.java:179)
> 	at it.auties.whatsapp.model.message.standard.ImageMessage$SimpleImageBuilder.create(ImageMessage.java:172)
> 	at be.geertdebaets.camera.service.WhatsAppService.sendImageMessageToGroup(WhatsAppService.java:147)
> 	... 1 more
> ```
> 
> Version: ` <version>3.0-RC13</version>` (but was also on RC12)
> 
> Is this a bug/problem, or i'm a doing something wrong in setting up the connections?

just fixed it, it will be available in the next update

**Comment by GDB4J** [comment]
> 



> I'm trying all types of connection examples but I can't get a QR code. When I try to do "api.connect().get()" (my last test is with some code of @GDB4J), but code is freezing and make serialized connection, under ".whatsappweb4j" only after stop application. I'm on MacOs.
> 
> There is some solutions?

Hi @19Maury85 ,
I've struggled aslo with getting a QRCode at first.

The easiest way to start is running one of the examples and see if you can get a QRCode there.
There you will see the easiest way to get a connection is this one (I believe):

```
var api = Whatsapp.lastConnection()
                .connect()
                .get();
```

It wil create a new connection if no connection was saved before.
Check also your {user.home}/.whatsappweb4j folder and perhaps try to delete everything from there
On windows C:/users/<your_username>/.whatsappweb4j
On Linux /home/<your_username>/.whatsappweb4j

My problem ended up using a too recent dependency version of "jakarta.websocket-api"
See issues (but this when working with Spring Boot) :
https://github.com/Auties00/WhatsappWeb4j/issues/111
https://github.com/Auties00/WhatsappWeb4j/issues/105

Kind regards

**Comment by 19Maury85** [comment]
I try all of the examples.
There is a possibility that I receive QR code but not printing to console?
because is seems to be freezed.
I already set logger to DEBUG on all library classes, to print into console

**Comment by GDB4J** [comment]
@19Maury85 
Normally it's not logged in any 'logger', so putting your logger to DEBUG (or ALL) has no effect.
The QRCode is default printed to the console (do a System.out.println("Test"); to see where is your console logging if you have no idea).
If you on Linux, check logfile /var/log/syslog

**Comment by Auties00** [comment]
> Reference i

it's weird, it should print it. What version are you using?

**Comment by 19Maury85** [comment]
the last one 3.0-RC13, but have this problem from previews versions (I'm started from first stable multidevice version)

**Comment by Auties00** [comment]
> the last one 3.0-RC13, but have this problem from previews versions (I'm started from first stable multidevice version)

Maybe it's your terminal? Can you try to set the qr handler to toFile instead of toTerminal inside Options?

**Comment by 19Maury85** [comment]
I need to implement a custom qr handler? because I use the default one

**Comment by 19Maury85** [comment]
Setting log to ALL, now is moving, but gone in error:

Caused by: javax.naming.NoInitialContextException: Need to specify class name in environment or system property, or in an application resource file: java.naming.factory.initial
	at java.naming/javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:704)
	at java.naming/javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305)
	at java.naming/javax.naming.InitialContext.getURLOrDefaultInitCtx(InitialContext.java:342)
	at java.naming/javax.naming.InitialContext.lookup(InitialContext.java:409)
	... 65 common frames omitted

and then:

java.io.IOException: connection closed locally

**Comment by Auties00** [comment]
> 

Are you running in some particular environment? 
Ex. Spring boot

**Comment by Auties00** [comment]
> I need to implement a custom qr handler? because I use the default one

Use QrHandler.toFile()

**Comment by 19Maury85** [comment]
> Are you running in some particular environment?
> Ex. Spring boot

Yes exactly Spring Boot.
If I try to run it in a main, I get another error (missing a util class that I already checked and it is present in the same package of class that get error)

**Comment by 19Maury85** [comment]
> > I need to implement a custom qr handler? because I use the default one
> 
> Use QrHandler.toFile()

Done but nothing...
I have the exception reported in preview message

**Comment by Auties00** [comment]
> > Are you running in some particular environment?
> > Ex. Spring boot
> 
> Yes exactly Spring Boot. If I try to run it in a main, I get another error (missing a util class that I already checked and it is present in the same package of class that get error)

It looks like it's a configuration problem on your end at this point. Can you share the full project in a GitHub repo so i can try to help you out?

**Comment by GDB4J** [comment]
Hi @19Maury85 

If you using spring boot (I'm on version 2.6.6):
Check my comments on 
https://github.com/Auties00/WhatsappWeb4j/issues/105

Add the correct dependency:

```
        <dependency>
            <groupId>com.github.auties00</groupId>
            <artifactId>whatsappweb4j</artifactId>
            <version>3.0-RC13</version>
        </dependency>
        <!-- IMPORTANT!  whatsappweb4j does not generate QR Codes when using version 2.1.0 of jakarta.websocket-api -->
        <dependency>
            <groupId>jakarta.websocket</groupId>
            <artifactId>jakarta.websocket-api</artifactId>
            <version>2.0.0</version>
        </dependency>
```

Are you running it on JDK 17.x? (not 18)

Also important, try to start your whats app connection after your spring boot application is fully started.
Put this in the main class of your spring boot application

```
@EventListener(ApplicationReadyEvent.class)
public void doSomethingAfterStartup() {
var api = Whatsapp.lastConnection()
                .connect()
                .get();
}
```

Check also your {user.home}/.whatsappweb4j folder and delete this folder first

That should normally do it, (at least it did for me ...)

But @Auties00 
We're getting way out of toppic here ...

Kind regards

**Comment by 19Maury85** [comment]
> > > Are you running in some particular environment?
> > > Ex. Spring boot
> > 
> > 
> > Yes exactly Spring Boot. If I try to run it in a main, I get another error (missing a util class that I already checked and it is present in the same package of class that get error)
> 
> It looks like it's a configuration problem on your end at this point. Can you share the full project in a GitHub repo so i can try to help you out?

Hi @Auties00 I sent you invitation.
Please ping me if there is access problem (is the first time I use github to share a my project).
Thanks in advance

**Comment by 19Maury85** [comment]
> GDB4J

@GDB4J  added you to (if you can help me)

**Comment by 19Maury85** [comment]
> Hi @19Maury85
> 
> If you using spring boot (I'm on version 2.6.6): Check my comments on #105
> 
> Add the correct dependency:
> 
> ```
>         <dependency>
>             <groupId>com.github.auties00</groupId>
>             <artifactId>whatsappweb4j</artifactId>
>             <version>3.0-RC13</version>
>         </dependency>
>         
>         <dependency>
>             <groupId>jakarta.websocket</groupId>
>             <artifactId>jakarta.websocket-api</artifactId>
>             <version>2.0.0</version>
>         </dependency>
> ```
> 
> Are you running it on JDK 17.x? (not 18)
> 
> Also important, try to start your whats app connection after your spring boot application is fully started. Put this in the main class of your spring boot application
> 
> ```
> @EventListener(ApplicationReadyEvent.class)
> public void doSomethingAfterStartup() {
> var api = Whatsapp.lastConnection()
>                 .connect()
>                 .get();
> }
> ```
> 
> Check also your {user.home}/.whatsappweb4j folder and delete this folder first
> 
> That should normally do it, (at least it did for me ...)
> 
> But @Auties00 We're getting way out of toppic here ...
> 
> Kind regards

I check all you points:

- dependency of jakarta was exactly 2.1.0 (change with your version)
- Jdk used is correctly 17
- Running connection into main class already done but get another error. Running into SpringBoot Application main class, not printing but seems to be not freezeed
- {user.home}/.whatsappweb4j already deleted more and more times
No Resolution....

About "We're getting way out of toppic here ..." is correct but I use this topic for 2 reasons: 

1. @Auties00 already answer you that your problem was fixed with new version
2. My problem is about connection to, so I use this topic to don't open new one. If you prefer I can make new one (tell me what you prefer).

Thanks in advance

**Comment by GDB4J** [comment]
@19Maury85 
I have access to your project, will look into it asap
At first sight I see a lot of things I mentioned, that is not the case on your project

I think it's better te create a new issue, because this is all about getting started in my eyes.
But @Auties00 is the admin here, he decide.
Anyway it is always helpfull for others too.

**Comment by Auties00** [comment]
> @19Maury85 I have access to your project, will look into it asap At first sight I see a lot of things I mentioned, that is not the case on your project
> 
> I think it's better te create a new issue, because this is all about getting started in my eyes. But @Auties00 is the admin here, he decide. Anyway it is always helpfull for others too.

if it isn't fixed rapidly we can definitely create a new issue. Btw I was looking into switching to the default implementation of WebSockets in Java introduced in Java 11 alongside the HTTP API, but I can't seem to make it work correctly. I can't tell if I'm doing something wrong or if it's a bug in the implementation(nobody seems to use it lol so it could be)

**Comment by Auties00** [comment]
If any more help is needed open a new issue

**Comment by Auties00** [comment]
>

**Comment by Auties00** [comment]
> Hi @Auties00
> 
> The root of this issue is'nt resolved in version `<version>3.0-RC14</version>`
> 
> Somehow creating a imagemessage is only possible when using a new connection, working on lastconnection gives the NPE below:
> 
> ```
> java.lang.NullPointerException: Cannot invoke "it.auties.whatsapp.model.media.MediaConnection.auth()" because the return value of "it.auties.whatsapp.controller.Store.mediaConnection()" is null
> 	at it.auties.whatsapp.util.Medias.upload(Medias.java:51)
> 	at it.auties.whatsapp.model.message.standard.ImageMessage.simpleBuilder(ImageMessage.java:179)
> 	at it.auties.whatsapp.model.message.standard.ImageMessage$SimpleImageBuilder.create(ImageMessage.java:172)
> ```
> 
> Can you reopen (or fix) this issue?
> 
> Thanks in advance, Kind regards,

Can you delete your old sessions and than retry? Because it works for me

**Comment by GDB4J** [comment]
@Auties00 
I did delete the old sessions. (<home_folder>/.whatsappweb4j)
created a new one and (restarted my application) then used the last connection

**Comment by Auties00** [comment]
I have pushed a new version called RC15 which should have fixed this issue completly

**Comment by Auties00** [comment]
> @Auties00 Is it possible version RC15 isn't available yet? Don't see a release commit & my maven can't get the dependency atm

Check the packages section, btw it should be available

**Comment by GDB4J** [comment]
> > @Auties00 Is it possible version RC15 isn't available yet? Don't see a release commit & my maven can't get the dependency atm
> 
> Check the packages section, btw it should be available


```
<dependency>
  <groupId>com.github.auties00</groupId>
  <artifactId>whatsappweb4j</artifactId>
  <version>3.0-RC15</version>
</dependency>
```


Gives me the result:

`Failed to read artifact descriptor for com.github.auties00:whatsappweb4j:jar:3.0-RC15`

"<version>3.0-RC14</version>" works well
what am I doing wrong here?

**Comment by Auties00** [comment]
> > > @Auties00 Is it possible version RC15 isn't available yet? Don't see a release commit & my maven can't get the dependency atm
> > 
> > 
> > Check the packages section, btw it should be available
> 
> ```
> <dependency>
>   <groupId>com.github.auties00</groupId>
>   <artifactId>whatsappweb4j</artifactId>
>   <version>3.0-RC15</version>
> </dependency>
> ```
> 
> Gives me the result:
> 
> `Failed to read artifact descriptor for com.github.auties00:whatsappweb4j:jar:3.0-RC15`
> 
> "3.0-RC14" works well what am I doing wrong here?

I've just pushed a new version, it may take some time for it to show up.
https://github.com/Auties00/WhatsappWeb4j/packages/1519983

**Comment by GDB4J** [comment]
> > > > @Auties00 Is it possible version RC15 isn't available yet? Don't see a release commit & my maven can't get the dependency atm
> > > 
> > > 
> > > Check the packages section, btw it should be available
> > 
> > 
> > ```
> > <dependency>
> >   <groupId>com.github.auties00</groupId>
> >   <artifactId>whatsappweb4j</artifactId>
> >   <version>3.0-RC15</version>
> > </dependency>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Gives me the result:
> > `Failed to read artifact descriptor for com.github.auties00:whatsappweb4j:jar:3.0-RC15`
> > "3.0-RC14" works well what am I doing wrong here?
> 
> I've just pushed a new version, it may take some time for it to show up. https://github.com/Auties00/WhatsappWeb4j/packages/1519983

doesn't work either.
I don't see the version in the tags:
https://github.com/Auties00/WhatsappWeb4j/tags
Is this normal?
All other versions works well (last versions 15 & 16 doesn't work for me (maven))

**Comment by Auties00** [comment]
> > > > > @Auties00 Is it possible version RC15 isn't available yet? Don't see a release commit & my maven can't get the dependency atm
> > > > 
> > > > 
> > > > Check the packages section, btw it should be available
> > > 
> > > 
> > > ```
> > > <dependency>
> > >   <groupId>com.github.auties00</groupId>
> > >   <artifactId>whatsappweb4j</artifactId>
> > >   <version>3.0-RC15</version>
> > > </dependency>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Gives me the result:
> > > `Failed to read artifact descriptor for com.github.auties00:whatsappweb4j:jar:3.0-RC15`
> > > "3.0-RC14" works well what am I doing wrong here?
> > 
> > 
> > I've just pushed a new version, it may take some time for it to show up. https://github.com/Auties00/WhatsappWeb4j/packages/1519983
> 
> doesn't work either. I don't see the version in the tags: https://github.com/Auties00/WhatsappWeb4j/tags Is this normal? All other versions works well (last versions 15 & 16 doesn't work for me (maven))

I was trying out github packages, looks like it doesn't work that well. I switched back. Soon you should see RC17 on mvn central. If it doesn't show up, open an issue

**Comment by GDB4J** [comment]
> > > > > > @Auties00 Is it possible version RC15 isn't available yet? Don't see a release commit & my maven can't get the dependency atm
> > > > > 
> > > > > 
> > > > > Check the packages section, btw it should be available
> > > > 
> > > > 
> > > > ```
> > > > <dependency>
> > > >   <groupId>com.github.auties00</groupId>
> > > >   <artifactId>whatsappweb4j</artifactId>
> > > >   <version>3.0-RC15</version>
> > > > </dependency>
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Gives me the result:
> > > > `Failed to read artifact descriptor for com.github.auties00:whatsappweb4j:jar:3.0-RC15`
> > > > "3.0-RC14" works well what am I doing wrong here?
> > > 
> > > 
> > > I've just pushed a new version, it may take some time for it to show up. https://github.com/Auties00/WhatsappWeb4j/packages/1519983
> > 
> > 
> > doesn't work either. I don't see the version in the tags: https://github.com/Auties00/WhatsappWeb4j/tags Is this normal? All other versions works well (last versions 15 & 16 doesn't work for me (maven))
> 
> I was trying out github packages, looks like it doesn't work that well. I switched back. Soon you should see RC17 on mvn central. If it doesn't show up, open an issue

@Auties00 

Still a no go:
```
        <dependency>
            <groupId>com.github.auties00</groupId>
            <artifactId>whatsappweb4j</artifactId>
            <version>3.0-RC17</version>
        </dependency>
```

results in:

**Failed to read artifact descriptor for com.github.auties00:link-preview:jar:1.3**
